content/4_Gold/PURS.mdx

```diff
@@ -95,6 +92,105 @@ using std::endl;
 using std::vector;
 
 /** A data structure that can answer point update & range minimum queries. */
+template <class T> class MinSegmentTree {
+  private:
+	const T DEFAULT = std::numeric_limits<T>().max();
+
+	int len;
+	int padded_len;
```

Comment:
oh btw does this need to be here

Reply:
wait im slightly confused - why are you setting at_right to be the padded length? are you making the array itself bigger to make it a perfect power of 2 or something?

Reply:
uh yeah

Reply:
> uh yeah

if ur goal is to minimize the memory usage, the typical way to do it is to keep the length of the array the same, but set the tree size to `4 << __lg(n)`, if I'm not mistaken

Reply:
uh how's that different from what i'm doing

Reply:
functionally it doesn’t matter, but you are effectively making the array ur querying on bigger

actually, it could be an issue if ur doing a O(n) segtree build because the array ur querying on is often bigger than n, if that makes sense

Reply:
i mean it only differs by a constant factor

Reply:
yeah but i think O(n) segtree build wouldn’t work with ur current code